hypothesis/client
# Add support for search sidebar panel via feature flag

This PR is an attempt to fix https://github.com/hypothesis/client/issues/5963

The issue described there is hard to fix without refactoring the whole search experience. This PR adds a new way to render the annotations search on a sidebar panel, which gives it more space and makes it easier to control when to display it.

This feature is handled via a feature flag, so that we can experiment with it, compare, and roll it back if it is not successful.

[Grabación de pantalla desde 2023-12-05 09-33-03.webm](https://github.com/hypothesis/client/assets/2719332/4a856e86-a753-4443-9d7b-7ecd1ac91d01)

### Implementation decisions

In order to make it easy to have both experiences at the same time and enable one or another via feature flag, the next decisions have been made.

* All existing components have been kept almost verbatim, with a few exceptions:
    1. The search input displayed in the stream page (http://localhost:5000/stream) has been replaced with the new `SearchField` component regardless of the feature flag value, as it's more accessible and will make it easier to finish the migration.
    2. Due to the point above, the old `SearchInput` has dropped the `alwaysExpanded` prop that was used for the stream section only.
* The feature flag is used only in a couple of places to dynamically render the search "entry points".
    1. If it's disabled, we continue rendering the existing `SearchInput` in the top bar.
    2. If it's enabled, we render the new `SearchIconButton` in the top bar. This is a toggle for the new `SearchPanel` component.
    3. The `SearchPanel` is appended to the sidebar only of the feature is enabled.
* When the feature is enabled, the new `SearchPanel` is displayed if the top bar toggle is active. This panel wraps the new `SearchField` and `SearchStatus`. The later includes the logic from `FilterStatus` related only with query filtering.
* The rest of the logic from `FilterStatus` has been put in a new `FilterAnnotationsStatus`, which has all the logic except for the one about filtering by query.
  This allows to continue displaying a standalone card when sharing one specific annotation or enabling focus mode, without having to display the search panel.
* The new `SearchInput` is a bit more accessible, by displaying an explicit submit button next to the search input.

All other improvements described in https://github.com/hypothesis/client/pull/5971 are also included here.

There's an unfortunate side effect on the decision to keep both new and old branches around with a feature flag, which is that there's some duplicated code, and this PR is a bit larger than it would be ideal.

### Testing steps

1. Check out this branch.
2. Check that all filtering and searching capabilities continue working as before.
3. Go to http://localhost:5000/admin/features and enable the `search_panel` feature flag.
4. Reload the page and check that the new searching experience via sidebar panel is enabled now.

### TODO

- [x] Add tests for new components
- [x] Make sure `/` character is not printed in search input when focus is recovered by pressing <kbd>/</kdb>
